File: android-app/src/main/kotlin/com/jworks/kanjiquest/android/ui/detail/RadicalDetailViewModel.kt
package com.jworks.kanjiquest.android.ui.detail

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.jworks.kanjiquest.core.domain.model.Kanji
import com.jworks.kanjiquest.core.domain.model.Radical
import com.jworks.kanjiquest.core.domain.repository.KanjiRepository
import com.jworks.kanjiquest.core.domain.repository.RadicalRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class RadicalDetailUiState(
    val radical: Radical? = null,
    val exampleKanji: List<Kanji> = emptyList(),
    val isLoading: Boolean = true
)

@HiltViewModel
class RadicalDetailViewModel @Inject constructor(
    savedStateHandle: SavedStateHandle,
    private val radicalRepository: RadicalRepository,
    private val kanjiRepository: KanjiRepository
) : ViewModel() {

    val radicalId: Int = checkNotNull(savedStateHandle["radicalId"])

    private val _uiState = MutableStateFlow(RadicalDetailUiState())
    val uiState: StateFlow<RadicalDetailUiState> = _uiState.asStateFlow()

    init {
        loadRadical()
    }

    private fun loadRadical() {
        viewModelScope.launch {
            try {
                val radical = radicalRepository.getRadicalById(radicalId)
                val kanjiIds = if (radical != null) {
                    radicalRepository.getKanjiIdsForRadical(radicalId)
                } else emptyList()

                val kanji = kanjiIds.mapNotNull { id ->
                    kanjiRepository.getKanjiById(id.toInt())
                }.sortedWith(compareBy<Kanji> { it.grade ?: 99 }.thenBy { it.frequency ?: 9999 })

                _uiState.value = RadicalDetailUiState(
                    radical = radical,
                    exampleKanji = kanji,
                    isLoading = false
                )
            } catch (e: Exception) {
                _uiState.value = RadicalDetailUiState(
                    radical = null,
                    exampleKanji = emptyList(),
                    isLoading = false
                )
            }
        }
    }
}

File: ios-app/KanjiQuest/Screens/Detail/RadicalDetailViewModel.swift
import Foundation
import SharedCore

@MainActor
final class RadicalDetailViewModel: ObservableObject {
    @Published var radical: Radical? = nil
    @Published var exampleKanji: [Kanji] = []
    @Published var isLoading = true

    private var radicalRepository: RadicalRepository?
    private var kanjiRepository: KanjiRepository?

    func load(container: AppContainer, radicalId: Int32) {
        radicalRepository = container.radicalRepository
        kanjiRepository = container.kanjiRepository

        Task {
            do {
                let rad = try await radicalRepository?.getRadicalById(id: radicalId)
                var kanji: [Kanji] = []
                if rad != nil {
                    let kanjiIds = (try? await radicalRepository?.getKanjiIdsForRadical(radicalId: radicalId)) ?? []
                    for id in kanjiIds {
                        if let k = try? await kanjiRepository?.getKanjiById(id: Int32(id)) {
                            kanji.append(k)
                        }
                    }
                    kanji.sort { lhs, rhs in
                        let lGrade = lhs.grade ?? 99
                        let rGrade = rhs.grade ?? 99
                        if lGrade != rGrade { return lGrade < rGrade }
                        let lFreq = lhs.frequency ?? 9999
                        let rFreq = rhs.frequency ?? 9999
                        return lFreq < rFreq
                    }
                }
                self.radical = rad
                self.exampleKanji = kanji
            } catch {
                self.radical = nil
                self.exampleKanji = []
            }
            self.isLoading = false
        }
    }
}
